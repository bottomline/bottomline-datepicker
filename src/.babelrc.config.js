const { declare } = require('@babel/helper-plugin-utils');

function getEnvOption(envOptions, env, preset) {
  return envOptions[env] ? envOptions[env][preset] : {};
}

module.exports = declare((api, presetOptions = {}) => {
  // see docs about api at https://babeljs.io/docs/en/config-files#apicache
  api.assertVersion('^7.0.0');

  const { env: envOptions = {}, ...options } = presetOptions;

  return {
    presets: [
      ['@babel/preset-env', {
        modules: false,
        ...options['@babel/preset-env']
      }],
      '@babel/preset-react'
    ],
    plugins: [
      '@babel/plugin-proposal-class-properties',
      ['@babel/plugin-proposal-object-rest-spread', {
        useBuiltIns: true,
        ...options['@babel/plugin-proposal-object-rest-spread']
      }]
    ],
    env: {
      test: {
        presets: [
          ['@babel/preset-env', {
            targets: {
              node: 'current'
            },
            ...getEnvOption(envOptions, 'test', '@babel/preset-env')
          }],
          '@babel/preset-react'
        ],
        plugins: [
          '@babel/plugin-proposal-class-properties',
          ['@babel/plugin-proposal-object-rest-spread', {
            useBuiltIns: true,
            ...getEnvOption(envOptions, 'test', '@babel/plugin-proposal-object-rest-spread')
          }]
        ]
      }
    }
  };
});
