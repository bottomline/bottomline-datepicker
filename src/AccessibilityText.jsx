import injectSheet from 'react-jss';
import PropTypes from 'prop-types';
import React from 'react';

const styles = {
  root: {
    position: 'absolute',
    width: 1,
    height: 1,
    margin: -1,
    padding: 0,
    overflow: 'hidden',
    clip: 'rect(0, 0, 0, 0)',
    border: 0
  }
};

const AccessibilityText = ({ classes, children, ...rest }) => (
  <p className={classes.root} {...rest}>{children}</p>
);

AccessibilityText.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  children: PropTypes.node.isRequired
};

export const AccessibilityTextBase = AccessibilityText;
export default injectSheet(styles)(AccessibilityText);
