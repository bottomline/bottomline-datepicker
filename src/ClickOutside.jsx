import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';

const ClickOutside = ({
  children, onClickOutside
}) => {
  const node = useRef();

  useEffect(() => {
    const handleKeyDown = e => {
      if (e.keyCode === 27) { // Escape
        onClickOutside();
      }
    };
    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.addEventListener('keydown', handleKeyDown);
    };
  }, [onClickOutside]);

  useEffect(() => {
    const handleClick = e => {
      if (e.target.contains(node.current)) onClickOutside();
    };
    window.addEventListener('resize', onClickOutside);
    document.addEventListener('click', handleClick);

    return () => {
      window.removeEventListener('resize', onClickOutside);
      document.removeEventListener('click', handleClick);
    };
  }, [onClickOutside]);

  return (
    <div ref={node}>{children}</div>
  );
};

ClickOutside.propTypes = {
  children: PropTypes.node.isRequired,
  inputEl: PropTypes.shape({}),
  onClickOutside: PropTypes.func.isRequired
};

ClickOutside.defaultProps = {
  inputEl: undefined
};

export default ClickOutside;
