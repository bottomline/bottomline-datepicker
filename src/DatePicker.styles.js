export default () => ({
  datepicker: {
    position: 'relative'
  }
});
