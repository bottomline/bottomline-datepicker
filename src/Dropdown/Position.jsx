import injectSheet from 'react-jss';
import PropTypes from 'prop-types';
import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import styles from './Position.styles';

const Position = ({
  children, classes, labelText, positionY, positionX, screenReaderLabel
}) => {
  let element = useRef();
  const [openDirection, setOpenDirection] = useState();
  const [dropDirection, setDropDirection] = useState();

  useEffect(() => {
    const {
      top, left, right, height
    } = element.getBoundingClientRect();
    const { innerWidth, innerHeight } = window;
    const bottomEdgePassesWindow = top + height > innerHeight;
    const spaceAboveIsGreater = top > innerHeight - top;
    const spaceLeftIsGreater = left > innerWidth - right;
    setOpenDirection(positionX);
    setDropDirection(positionY);

    if (spaceLeftIsGreater) {
      setOpenDirection('left');
    }

    if (bottomEdgePassesWindow && spaceAboveIsGreater) {
      setDropDirection('up');
    }
  }, [positionX, positionY]);

  return (
    <div
      data-qa="dropdown-position"
      ref={node => { element = node; }}
      className={classNames({
        [classes.dropdown]: true,
        [classes.dropdownLeft]: openDirection === 'left',
        [classes.dropdownCenter]: openDirection === 'center',
        [classes.dropdownUp]: dropDirection === 'up',
        [classes.screenReaderLabel]: !labelText || screenReaderLabel
      })}
    >
      {children}
    </div>
  );
};

Position.propTypes = {
  children: PropTypes.node.isRequired,
  classes: PropTypes.shape({}).isRequired,
  labelText: PropTypes.string,
  positionX: PropTypes.string,
  positionY: PropTypes.string,
  screenReaderLabel: PropTypes.bool.isRequired
};

Position.defaultProps = {
  positionY: 'right',
  positionX: 'down',
  labelText: 'undefined'
};

export default injectSheet(styles)(Position);
