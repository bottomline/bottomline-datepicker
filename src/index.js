export { default as DatePicker, BareDatePicker } from './DatePicker';

// eslint-disable-next-line camelcase
export { default as defaultStrings_en_US } from './locale/defaultStrings_en_US';
