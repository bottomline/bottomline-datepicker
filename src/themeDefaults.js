/* istanbul ignore file */
import { registerDefaultsForModule } from '@glu/theming';
import legacy from './themes/legacyTheme';
import ash from './themes/ashTheme';

const themes = { ash, legacy };

registerDefaultsForModule(
  'datepickerReact', ({ palette, typography, name }) => themes[name](palette, typography)
);
