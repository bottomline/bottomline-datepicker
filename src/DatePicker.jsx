import { withStyles, withTheme } from '@glu/theming';
import moment from 'moment';
import PropTypes from 'prop-types';
import React, { useState } from 'react';
import Dropdown from './Dropdown';
import TextInput from './TextInput';
import ClickOutside from './ClickOutside';
import { datesToString } from './utils/formatString.js';
import styles from './DatePicker.styles';
import './themeDefaults';

const DatePicker = ({
  classes,
  className,
  dataQa,
  dateEnd,
  dateStart,
  htmlId,
  name,
  onChangeCallback,
  screenReaderLabel,
  ...rest
}) => {
  const [open, setOpen] = useState(false);
  const [selectedStartDate, setSelectedStartDate] = useState(dateStart
    ? moment(dateStart)
    : undefined);
  const [selectedEndDate, setSelectedEndDate] = useState(dateEnd
    ? moment(dateEnd)
    : undefined);
  const [confirmedStartDate, setConfirmedStartDate] = useState(selectedStartDate);
  const [confirmedEndDate, setConfirmedEndDate] = useState(selectedEndDate);
  const [head, setHead] = useState(selectedStartDate ? moment(selectedStartDate) : moment());
  const [inputValue, setInputValue] = useState(
    datesToString(selectedStartDate, selectedEndDate, rest)
  );

  const handleApply = (start, end) => {
    setOpen(false);
    setConfirmedStartDate(start);
    setConfirmedEndDate(end);
    onChangeCallback({
      value: {
        date: moment(start),
        dateEnd: moment(end)
      }
    });
    setInputValue(datesToString(start, end, rest));
  };

  const handleClose = () => {
    setOpen(false);
    setSelectedStartDate(confirmedStartDate);
    setSelectedEndDate(confirmedEndDate);
  };

  return (
    <ClickOutside onClickOutside={handleClose}>
      <div data-qa={dataQa ? `${dataQa}` : `datepicker-${htmlId}`} className={`${classes.datepicker} ${className}`}>
        <TextInput {...{
          handleApply,
          handleClose,
          htmlId,
          inputValue,
          name,
          open,
          screenReaderLabel,
          selectedEndDate,
          selectedStartDate,
          setHead,
          setInputValue,
          setOpen,
          setSelectedEndDate,
          setSelectedStartDate,
          ...rest
        }}
        />
        {open && (

        <Dropdown {...{
          head,
          onApply: handleApply,
          onClose: handleClose,
          screenReaderLabel,
          selectedEndDate,
          selectedStartDate,
          setHead,
          setSelectedEndDate,
          setSelectedStartDate,
          ...rest
        }}
        />
        )}
      </div>
    </ClickOutside>
  );
};

DatePicker.propTypes = {
  /** Toggle whether to always show calendar when a range is selected. */
  alwaysShowCalendars: PropTypes.bool,

  /** Custom classes for the apply button */
  applyButtonClasses: PropTypes.string,

  /** Hide the apply and cancel buttons, and automatically apply a new selection */
  autoApply: PropTypes.bool,

  /** Custom classes for container that wraps action buttons */
  buttonClasses: PropTypes.string,

  /** Custom classes for the cancel button */
  cancelButtonClasses: PropTypes.string,

  /** Classes provided by the theme */
  classes: PropTypes.shape({}).isRequired,

  /** Class passed to parent element for extra decoration. */
  className: PropTypes.string,

  /** String given to parent elements data-qa html attribute for easy targeting */
  dataQa: PropTypes.string,

  /** The end of initial selected date range. Must be ISO format 'YYYY-MM-DD' */
  dateEnd: PropTypes.string,

  /** The latest date a user may select. Must be ISO format 'YYYY-MM-DD' */
  dateMax: PropTypes.string,

  /** The earliest date a user may select. Must be ISO format 'YYYY-MM-DD' */
  dateMin: PropTypes.string,

  /** The start of initial selected date range. Must be ISO format 'YYYY-MM-DD' */
  dateStart: PropTypes.string,

  /** Error message. will add styles to input if present. */
  error: PropTypes.string,

  /** Format to display date string in input */
  format: PropTypes.string,

  /** Function which is passed each visible day and returns string of classNames. */
  getClassNamesForDate: PropTypes.func,

  /** string assigned to HTML id attribute of component */
  htmlId: PropTypes.string,

  /* Function which is passed each visible date and returns boolean for if it is selectable. */
  isSelectableDate: PropTypes.func,

  /** Display weeks as ISO week numbers; weekNumbers must also be true. */
  isoWeekNumbers: PropTypes.bool,

  /** String used for label value */
  labelText: PropTypes.string,

  /** Toggle to control calendar navigation. Defaults to true. Can be set to
   * false to navigate calendars independently. */
  linkedCalendars: PropTypes.bool,

  /** Show select dropdown for the Month in the calendar header */
  monthDropdown: PropTypes.bool,

  /** string passed to html name attribute of text input */
  name: PropTypes.string,

  /** The number of calendars to display side by side. */
  numberOfMonths: PropTypes.number,

  /** Function that triggers once a date is selected or changed */
  onChangeCallback: PropTypes.func,

  /** String passed as html placeholder text for input. */
  placeholder: PropTypes.string,

  /**
  * Whether the picker appears below (default) or above the HTML element it's attached to.
  * ("down"/"up") */
  positionY: PropTypes.string,

  /**
   * Whether the picker appears aligned to the left, to the right, or centered
   * ("left"/"right"/"center") */
  positionX: PropTypes.string,

  /** Delimiter used to separate date string in text input */
  rangeDelimiter: PropTypes.string,

  /** Object with list of predefined ranges */
  rangePresets: PropTypes.shape({}),

  /** Toggle range preset feature */
  ranges: PropTypes.bool,

  /** Toggle to hide the label for all but screenreaders */
  screenReaderLabel: PropTypes.bool,

  /** Show "custom range" option to display calendar */
  showCustomRangeLabel: PropTypes.bool,

  /** Interacting with calendar sets a single date. */
  singleDatePicker: PropTypes.bool,

  /** Show timePicker */
  timePicker: PropTypes.bool,

  /** Display 24 hour military style time. Also hides am/pm selector. */
  timePicker24Hour: PropTypes.bool,

  /** Size of increments for minute selector. */
  timePickerIncrement: PropTypes.number,

  /** Show two timepickers to define a time range */
  timePickerRange: PropTypes.bool,

  /** Show seconds selector. timePicker or timePickerRange must be set */
  timePickerSeconds: PropTypes.bool,

  /** Display column left of calendar with week numbers */
  weekNumbers: PropTypes.bool,

  /** Show select dropdown for the Year in the calendar header */
  yearDropdown: PropTypes.bool,

  /** The maximum year in the select year dropdown */
  yearDropdownMaxYear: PropTypes.number,

  /** The minimum year in the select year dropdown */
  yearDropdownMinYear: PropTypes.number
};

DatePicker.defaultProps = {
  alwaysShowCalendars: false,
  applyButtonClasses: '',
  autoApply: false,
  buttonClasses: '',
  cancelButtonClasses: '',
  className: '',
  dataQa: '',
  dateEnd: undefined,
  dateMax: moment().add(50, 'y').format('YYYY-MM-DD'),
  dateMin: moment().add(-50, 'y').format('YYYY-MM-DD'),
  dateStart: undefined,
  error: '',
  format: 'D MMM YYYY',
  getClassNamesForDate: () => '',
  htmlId: '',
  isSelectableDate: () => false,
  isoWeekNumbers: false,
  labelText: '',
  linkedCalendars: true,
  monthDropdown: false,
  name: '',
  numberOfMonths: 1,
  onChangeCallback: () => {},
  placeholder: '',
  positionY: 'down',
  positionX: 'right',
  rangeDelimiter: 'â€“',
  ranges: false,
  rangePresets: {
    Today: [
      moment().format('YYYY-MM-DD'),
      moment().format('YYYY-MM-DD')
    ],
    Yesterday: [
      moment().subtract(1, 'day').format('YYYY-MM-DD'),
      moment().subtract(1, 'day').format('YYYY-MM-DD')
    ],
    'Last 7 Days': [
      moment().subtract(7, 'day').format('YYYY-MM-DD'),
      moment().format('YYYY-MM-DD')
    ],
    'Last 30 Days': [
      moment().subtract(30, 'day').format('YYYY-MM-DD'),
      moment().format('YYYY-MM-DD')
    ],
    'This Month': [
      moment().startOf('month').format('YYYY-MM-DD'),
      moment().format('YYYY-MM-DD')
    ],
    'Last Month': [
      moment().subtract(1, 'month').startOf('month').format('YYYY-MM-DD'),
      moment().subtract(1, 'month').endOf('month').format('YYYY-MM-DD')
    ]
  },
  screenReaderLabel: false,
  showCustomRangeLabel: false,
  singleDatePicker: false,
  timePicker: false,
  timePicker24Hour: false,
  timePickerIncrement: 1,
  timePickerRange: false,
  timePickerSeconds: false,
  weekNumbers: false,
  yearDropdown: false,
  yearDropdownMaxYear: parseInt(moment().add(50, 'y').format('YYYY'), 10),
  yearDropdownMinYear: parseInt(moment().add(-50, 'y').format('YYYY'), 10)
};

export const BareDatePicker = DatePicker;
export default withTheme(withStyles(styles)(DatePicker));
