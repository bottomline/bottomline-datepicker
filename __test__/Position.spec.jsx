import React from 'react';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';
import { nextTheme, ThemeProvider } from '@glu/theming';
import Position from '../src/Dropdown/Position';
import '../src/themeDefaults';

const Wrapper = ({ children }) => (
  <ThemeProvider baseTheme={nextTheme}>
    {children}
  </ThemeProvider>
);

Wrapper.propTypes = {
  children: PropTypes.node.isRequired
};

describe('Position', () => {
  afterEach(() => {
    document.body.innerHTML = '';
  });

  it('should change position based on window', () => {
    window.innerHeight = 500;
    window.innerWidth = 600;
    Element.prototype.getBoundingClientRect = jest.fn(() => ({
      width: 120,
      height: 120,
      top: 500,
      left: -100,
      bottom: 0,
      right: 0
    }));
    const wrapper = mount(
      <Wrapper>
        <div style={{
          position: 'absolute',
          bottom: 0,
          left: 0,
          top: 'auto',
          right: 'auto'
        }}
        >
          <Position screenReaderLabel>
            <div>Children</div>
          </Position>
        </div>
      </Wrapper>
    );
    expect(wrapper.find('[data-qa="dropdown-position"]')).toMatchSnapshot();
  });

  it('should change position when spaceLeftIsGreater', () => {
    window.innerHeight = 500;
    window.innerWidth = 600;
    Element.prototype.getBoundingClientRect = jest.fn(() => ({
      width: 120,
      height: 120,
      top: 500,
      left: 700,
      bottom: 0,
      right: 0
    }));
    const wrapper = mount(
      <Wrapper>
        <div style={{
          position: 'absolute',
          bottom: 0,
          left: 0,
          top: 'auto',
          right: 'auto'
        }}
        >
          <Position screenReaderLabel>
            <div>Children</div>
          </Position>
        </div>
      </Wrapper>
    );
    expect(wrapper.find('[data-qa="dropdown-position"]')).toMatchSnapshot();
  });
});
