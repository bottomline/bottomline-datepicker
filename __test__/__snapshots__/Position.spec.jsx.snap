// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Position should change position based on window 1`] = `
<div
  className="Position-dropdown-0-1-1 Position-dropdownUp-0-1-6 Position-screenReaderLabel-0-1-2"
  data-qa="dropdown-position"
>
  <div>
    Children
  </div>
</div>
`;

exports[`Position should change position when spaceLeftIsGreater 1`] = `
<div
  className="Position-dropdown-0-1-1 Position-dropdownLeft-0-1-4 Position-dropdownUp-0-1-6 Position-screenReaderLabel-0-1-2"
  data-qa="dropdown-position"
>
  <div>
    Children
  </div>
</div>
`;
